File: mobile/ios/Data/CacheManager.swift
import Foundation

class CacheManager {
    static let shared = CacheManager()
    
    private let memoryCache = NSCache<NSString, AnyObject>()
    private let fileManager = FileManager.default
    private let cacheDirectory: URL
    
    private init() {
        // Настройка кэш директории
        let paths = fileManager.urls(for: .cachesDirectory, in: .userDomainMask)
        cacheDirectory = paths[0].appendingPathComponent("ALADDINCache")
        
        // Создаем директорию если не существует
        try? fileManager.createDirectory(at: cacheDirectory, withIntermediateDirectories: true)
        
        // Настройка memory cache
        memoryCache.countLimit = 100
        memoryCache.totalCostLimit = 50 * 1024 * 1024 // 50 MB
    }
    
    // MARK: - Memory Cache
    
    func cacheInMemory(_ object: AnyObject, forKey key: String) {
        memoryCache.setObject(object, forKey: key as NSString)
        print("✅ Cached in memory: \(key)")
    }
    
    func getFromMemory(forKey key: String) -> AnyObject? {
        return memoryCache.object(forKey: key as NSString)
    }
    
    func removeFromMemory(forKey key: String) {
        memoryCache.removeObject(forKey: key as NSString)
    }
    
    func clearMemoryCache() {
        memoryCache.removeAllObjects()
        print("✅ Memory cache cleared")
    }
    
    // MARK: - Disk Cache
    
    func cacheToDisk<T: Codable>(_ object: T, forKey key: String) {
        let fileURL = cacheDirectory.appendingPathComponent(key)
        
        do {
            let data = try JSONEncoder().encode(object)
            try data.write(to: fileURL)
            print("✅ Cached to disk: \(key)")
        } catch {
            print("❌ Disk cache error: \(error.localizedDescription)")
        }
    }
    
    func getFromDisk<T: Codable>(forKey key: String, as type: T.Type) -> T? {
        let fileURL = cacheDirectory.appendingPathComponent(key)
        
        guard fileManager.fileExists(atPath: fileURL.path) else {
            return nil
        }
        
        do {
            let data = try Data(contentsOf: fileURL)
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            print("❌ Disk read error: \(error.localizedDescription)")
            return nil
        }
    }
    
    func removeFromDisk(forKey key: String) {
        let fileURL = cacheDirectory.appendingPathComponent(key)
        try? fileManager.removeItem(at: fileURL)
    }
    
    func clearDiskCache() {
        try? fileManager.removeItem(at: cacheDirectory)
        try? fileManager.createDirectory(at: cacheDirectory, withIntermediateDirectories: true)
        print("✅ Disk cache cleared")
    }
    
    // MARK: - Hybrid Cache (Memory + Disk)
    
    func cache<T: Codable>(_ object: T, forKey key: String) {
        // Memory cache
        cacheInMemory(object as AnyObject, forKey: key)
        
        // Disk cache
        cacheToDisk(object, forKey: key)
    }
    
    func get<T: Codable>(forKey key: String, as type: T.Type) -> T? {
        // Try memory cache first
        if let cachedObject = getFromMemory(forKey: key) as? T {
            print("✅ Hit memory cache: \(key)")
            return cachedObject
        }
        
        // Try disk cache
        if let diskObject = getFromDisk(forKey: key, as: type) {
            print("✅ Hit disk cache: \(key)")
            // Cache to memory for faster access
            cacheInMemory(diskObject as AnyObject, forKey: key)
            return diskObject
        }
        
        print("❌ Cache miss: \(key)")
        return nil
    }
    
    // MARK: - Cache Management
    
    func getCacheSize() -> Int64 {
        var totalSize: Int64 = 0
        
        guard let enumerator = fileManager.enumerator(at: cacheDirectory, includingPropertiesForKeys: [.fileSizeKey]) else {
            return 0
        }
        
        for case let fileURL as URL in enumerator {
            if let fileSize = try? fileURL.resourceValues(forKeys: [.fileSizeKey]).fileSize {
                totalSize += Int64(fileSize)
            }
        }
        
        return totalSize
    }
    
    func clearAllCache() {
        clearMemoryCache()
        clearDiskCache()
        print("✅ All cache cleared")
    }
    
    func cleanupOldCache(olderThan days: Int) {
        let cutoffDate = Date().addingTimeInterval(-TimeInterval(days * 24 * 60 * 60))
        
        guard let enumerator = fileManager.enumerator(at: cacheDirectory, includingPropertiesForKeys: [.contentModificationDateKey]) else {
            return
        }
        
        for case let fileURL as URL in enumerator {
            if let modificationDate = try? fileURL.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate,
               modificationDate < cutoffDate {
                try? fileManager.removeItem(at: fileURL)
            }
        }
        
        print("✅ Old cache cleaned up")
    }
}

File: mobile/android/Data/CacheManager.kt
package com.aladdin.mobile.data

import android.content.Context
import android.util.LruCache
import com.google.gson.Gson
import java.io.File
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class CacheManager @Inject constructor(
    private val context: Context
) {
    
    private val memoryCache: LruCache<String, Any>
    private val gson = Gson()
    private val cacheDirectory: File
    
    init {
        // Настройка memory cache (20% от доступной памяти)
        val maxMemory = (Runtime.getRuntime().maxMemory() / 1024).toInt()
        val cacheSize = maxMemory / 5
        
        memoryCache = LruCache(cacheSize)
        
        // Настройка кэш директории
        cacheDirectory = File(context.cacheDir, "ALADDINCache")
        if (!cacheDirectory.exists()) {
            cacheDirectory.mkdirs()
        }
        
        android.util.Log.i("CacheManager", "Initialized with cache size: ${cacheSize}KB")
    }
    
    // MARK: - Memory Cache
    
    fun cacheInMemory(key: String, value: Any) {
        memoryCache.put(key, value)
        android.util.Log.i("CacheManager", "Cached in memory: $key")
    }
    
    fun getFromMemory(key: String): Any? {
        return memoryCache.get(key)
    }
    
    fun removeFromMemory(key: String) {
        memoryCache.remove(key)
    }
    
    fun clearMemoryCache() {
        memoryCache.evictAll()
        android.util.Log.i("CacheManager", "Memory cache cleared")
    }
    
    // MARK: - Disk Cache
    
    fun <T> cacheToDisk(key: String, value: T) {
        val file = File(cacheDirectory, key)
        
        try {
            val json = gson.toJson(value)
            file.writeText(json)
            android.util.Log.i("CacheManager", "Cached to disk: $key")
        } catch (e: IOException) {
            android.util.Log.e("CacheManager", "Disk cache error", e)
        }
    }
    
    fun <T> getFromDisk(key: String, classOfT: Class<T>): T? {
        val file = File(cacheDirectory, key)
        
        if (!file.exists()) {
            return null
        }
        
        return try {
            val json = file.readText()
            gson.fromJson(json, classOfT)
        } catch (e: IOException) {
            android.util.Log.e("CacheManager", "Disk read error", e)
            null
        }
    }
    
    fun removeFromDisk(key: String) {
        val file = File(cacheDirectory, key)
        if (file.exists()) {
            file.delete()
        }
    }
    
    fun clearDiskCache() {
        cacheDirectory.deleteRecursively()
        cacheDirectory.mkdirs()
        android.util.Log.i("CacheManager", "Disk cache cleared")
    }
    
    // MARK: - Hybrid Cache (Memory + Disk)
    
    fun <T> cache(key: String, value: T) {
        // Memory cache
        cacheInMemory(key, value as Any)
        
        // Disk cache
        cacheToDisk(key, value)
    }
    
    fun <T> get(key: String, classOfT: Class<T>): T? {
        // Try memory cache first
        val memoryValue = getFromMemory(key)
        if (memoryValue != null && classOfT.isInstance(memoryValue)) {
            android.util.Log.i("CacheManager", "Hit memory cache: $key")
            return classOfT.cast(memoryValue)
        }
        
        // Try disk cache
        val diskValue = getFromDisk(key, classOfT)
        if (diskValue != null) {
            android.util.Log.i("CacheManager", "Hit disk cache: $key")
            // Cache to memory for faster access
            cacheInMemory(key, diskValue as Any)
            return diskValue
        }
        
        android.util.Log.i("CacheManager", "Cache miss: $key")
        return null
    }
    
    fun remove(key: String) {
        removeFromMemory(key)
        removeFromDisk(key)
    }
    
    // MARK: - Cache Management
    
    fun getCacheSize(): Long {
        return cacheDirectory.walkTopDown()
            .filter { it.isFile }
            .map { it.length() }
            .sum()
    }
    
    fun clearAllCache() {
        clearMemoryCache()
        clearDiskCache()
        android.util.Log.i("CacheManager", "All cache cleared")
    }
    
    fun cleanupOldCache(olderThanDays: Int) {
        val cutoffTime = System.currentTimeMillis() - (olderThanDays * 24 * 60 * 60 * 1000L)
        
        cacheDirectory.walkTopDown()
            .filter { it.isFile && it.lastModified() < cutoffTime }
            .forEach { it.delete() }
        
        android.util.Log.i("CacheManager", "Old cache cleaned up")
    }
    
    // MARK: - Cache Statistics
    
    fun getCacheStatistics(): CacheStatistics {
        val totalFiles = cacheDirectory.walkTopDown().filter { it.isFile }.count()
        val totalSize = getCacheSize()
        val memoryCacheSize = memoryCache.size()
        
        return CacheStatistics(
            totalFiles = totalFiles,
            totalSizeBytes = totalSize,
            totalSizeMB = totalSize / (1024 * 1024),
            memoryCacheEntries = memoryCacheSize
        )
    }
}

// MARK: - Data Models

data class CacheStatistics(
    val totalFiles: Int,
    val totalSizeBytes: Long,
    val totalSizeMB: Long,
    val memoryCacheEntries: Int
)

